### ISSUE TITLE
Bug: inconsistent output with transformer models between CUDA and CPU execution providers

### ISSUE THREAD

**Issue by rene-wilbers-adesso** [microsoft/onnxruntime#24204]
### Describe the issue

I created an ONNX graph of MobileSAM, a transformer model for segmentation. I created the combined encoder-decoder graph using the following code:

Shell:
```
mkdir weights
python3 -m venv venv
source venv/bin/activate
python3 -m pip install -r requirements.txt
git clone https://github.com/ChaoningZhang/MobileSAM.git
cd MobileSAM
python3 scripts/export_onnx_model.py --checkpoint ./weights/mobile_sam.pt --model-type vit_t --output ../weights/mobile_sam_decoder.onnx
cd ..   
python3 create_onnx.py
```

create_onnx.py:
```
import torch
import numpy as np
import os
from segment_anything.utils.transforms import ResizeLongestSide
from onnxruntime.quantization import QuantType
from onnxruntime.quantization.quantize import quantize_dynamic
import mobile_sam as SAM
from dataclasses import dataclass, field
from typing import Tuple
import onnx
from onnx import helper
from onnx import version_converter
from onnxconverter_common import float16
import sys
import os

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from mobilesam_endtoend import MobileSAM_endtoend


@dataclass
class Config:
    output_names: list[str] = field(default_factory=lambda: ["output"])
    checkpoint: str = "MobileSAM/weights/mobile_sam.pt"
    model_type: str = "vit_t"
    image_size: Tuple[int, int] = (1024, 720)
    opset: int = 19
    encoder_path: str = "weights/mobile_sam_encoder.onnx"
    decoder_path: str = "weights/mobile_sam_decoder.onnx"
    output_path: str = "weights/unwell_mobile_sam.2.onnx"
    test_input: str = "../testdata/frame1.png"
    test_bb: list[float] = field(default_factory=lambda: [310.5, 494.625, 432, 642.375])


class Model(torch.nn.Module):
    def __init__(self, config: Config):
        super().__init__()
        self.sam = SAM.sam_model_registry[config.model_type](
            checkpoint=config.checkpoint
        )
        self.sam.to(device="cpu")
        self.predictor = SAM.SamPredictor(self.sam)
        self.config = config

    def forward(self, x):
        self.predictor.set_torch_image(x, (self.config.image_size))
        if "interm_embeddings" not in self.config.output_names:
            return self.predictor.get_image_embedding()
        else:
            return self.predictor.get_image_embedding(), torch.stack(
                self.predictor.interm_features, dim=0
            )


def export_encoder(config: Config):
    os.makedirs(os.path.dirname(config.encoder_path), exist_ok=True)

    sam = SAM.sam_model_registry[config.model_type](checkpoint=config.checkpoint)
    sam.to(device="cpu")
    transform = ResizeLongestSide(sam.image_encoder.img_size)

    image = np.zeros((config.image_size[1], config.image_size[0], 3), dtype=np.uint8)
    input_image = transform.apply_image(image)
    input_image_torch = torch.as_tensor(input_image, device="cpu")
    input_image_torch = input_image_torch.permute(2, 0, 1).contiguous()[None, :, :, :]

    model = Model(config)
    model_trace = torch.jit.trace(model, input_image_torch)
    torch.onnx.export(
        model_trace,
        input_image_torch,
        config.encoder_path,
        input_names=["input"],
        output_names=config.output_names,
    )


def combine_encoder_decoder(config: Config):
    encoder_model = onnx.load(config.encoder_path)
    decoder_model = onnx.load(config.decoder_path)
    encoder_model = version_converter.convert_version(encoder_model, config.opset)
    decoder_model = version_converter.convert_version(decoder_model, config.opset)

    # Rename encoder output to match decoder input
    encoder_model.graph.output[0].name = "image_embeddings"
    encoder_model.graph.node[-1].output[0] = "image_embeddings"

    # Get encoder inputs and outputs
    encoder_inputs = [node.name for node in encoder_model.graph.input]
    encoder_outputs = [node.name for node in encoder_model.graph.output]

    # Get decoder inputs and outputs
    decoder_inputs = [node.name for node in decoder_model.graph.input]
    decoder_outputs = [node.name for node in decoder_model.graph.output]

    print("Encoder inputs:", encoder_inputs)
    print("Encoder outputs:", encoder_outputs)
    print("Decoder inputs:", decoder_inputs)
    print("Decoder outputs:", decoder_outputs)

    # Add function to rename nodes with a prefix
    def add_prefix_to_nodes(model, prefix):
        # Create a mapping of old names to new names
        name_mapping = {}

        # First pass: collect all output names that need to be renamed
        for node in model.graph.node:
            for output in node.output:
                # Don't rename the image_embeddings output as it needs to connect to the decoder
                if (
                    output not in ["image_embeddings"]
                    and output not in decoder_inputs
                    and output not in decoder_outputs
                ):
                    name_mapping[output] = prefix + output

        # Second pass: rename nodes and their connections
        for node in model.graph.node:
            node.name = prefix + node.name

            # Rename outputs
            for i, output in enumerate(node.output):
                if output in name_mapping:
                    node.output[i] = name_mapping[output]

            # Rename inputs (but only if they're in our mapping)
            for i, input_name in enumerate(node.input):
                if input_name in name_mapping:
                    node.input[i] = name_mapping[input_name]

        return model

    # Add this before combining the models
    encoder_model = add_prefix_to_nodes(encoder_model, "encoder_")
    first_encoder_node = next(
        node for node in encoder_model.graph.node if "input" in node.input
    )
    # Add a transpose operation to the input named 'input'
    transpose_node = onnx.helper.make_node(
        "Transpose", inputs=["input"], outputs=["transposed_input"], perm=[0, 3, 1, 2]
    )
    # Insert the transpose node into the encoder model's graph
    encoder_model.graph.node.insert(0, transpose_node)
    # Update the input shape of the model
    input_shape = encoder_model.graph.input[0].type.tensor_type.shape
    # Change dimensions to [batch_size, height, width, channels]
    input_shape.dim[0].dim_value = 1  # batch size
    input_shape.dim[1].dim_value = 720  # height
    input_shape.dim[2].dim_value = 1024  # width
    input_shape.dim[3].dim_value = 3  # channels (RGB)

    # Update the input name for the first node in the encoder model to use the transposed input
    for i, input_name in enumerate(first_encoder_node.input):
        if input_name == "input":
            first_encoder_node.input[i] = "transposed_input"

    combined_inputs = list(encoder_model.graph.input) + [
        node for node in decoder_model.graph.input if node.name != "image_embeddings"
    ]

    # Create constant tensor for threshold value (0)
    threshold_const = helper.make_tensor(
        name="threshold_value",
        data_type=onnx.TensorProto.FLOAT16,  # Match input type
        dims=[1],
        vals=[0.0],
    )

    # Create Greater node for thresholding
    threshold_node = helper.make_node(
        "Greater",
        inputs=["masks", "threshold_value"],
        outputs=["thresholded_masks"],
        name="threshold_masks",
    )

    # Add Cast node to convert boolean to float16
    bool_to_float_node = helper.make_node(
        "Cast",
        inputs=["thresholded_masks"],
        outputs=["float_masks"],
        to=onnx.TensorProto.FLOAT16,
        name="cast_to_float16",
    )

    # Create constant tensor for axis
    axis_const = helper.make_tensor(
        name="axis_value", data_type=onnx.TensorProto.INT64, dims=[1], vals=[1]
    )

    # Update ReduceMax to use float input
    reduce_node = helper.make_node(
        "ReduceMax",
        inputs=["float_masks", "axis_value"],
        outputs=["reduced_masks"],
        keepdims=0,
        name="reduce_masks",
    )

    # Multiply by 255
    multiply_const = helper.make_tensor(
        name="multiply_value", data_type=onnx.TensorProto.FLOAT, dims=[1], vals=[255.0]
    )

    multiply_node = helper.make_node(
        "Mul",
        inputs=["reduced_masks", "multiply_value"],
        outputs=["foreground_mask"],
        name="multiply_by_255",
    )

    # Add Cast node to convert float16 to uint8
    float_to_uint8_node = helper.make_node(
        "Cast",
        inputs=["foreground_mask"],
        outputs=["uint8_mask"],
        to=onnx.TensorProto.UINT8,
        name="cast_to_uint8",
    )

    # Update combined nodes list to include the new cast node
    combined_nodes = (
        list(encoder_model.graph.node)
        + list(decoder_model.graph.node)
        + [
            threshold_node,
            bool_to_float_node,
            reduce_node,
            multiply_node,
            float_to_uint8_node,
        ]
    )

    # Update combined initializers to include all constants
    combined_initializers = (
        list(encoder_model.graph.initializer)
        + list(decoder_model.graph.initializer)
        + [multiply_const, threshold_const, axis_const]
    )

    # Update the graph output to use our final mask
    new_output = helper.make_tensor_value_info(
        "uint8_mask",
        onnx.TensorProto.UINT8,
        None,  # Shape will be inferred
    )

    # Modify the graph creation to use the new output
    combined_graph = helper.make_graph(
        combined_nodes,
        "encoder_decoder_model",
        combined_inputs,
        [new_output],  # Replace the original outputs with our new output
        combined_initializers,
    )

    print("Number of nodes:", len(combined_graph.node))

    combined_model = helper.make_model(
        combined_graph,
        producer_name="encoder_decoder_combined",
        opset_imports=[helper.make_operatorsetid("", config.opset)],
    )

    model_fp16 = float16.convert_float_to_float16(combined_model)
    onnx.save(model_fp16, config.output_path)


def test_model(config: Config, provider: str = "CUDAExecutionProvider"):
    model = MobileSAM_endtoend(config.output_path, providers=[provider])
    out = model.get_foreground_mask(
        config.test_input, np.array(config.test_bb, dtype=np.float16)
    )
    assert out.sum() > 0
    assert out.dtype == np.uint8


if __name__ == "__main__":
    config = Config()
    export_encoder(config)
    combine_encoder_decoder(config)
    try:
        test_model(config, provider="CPUExecutionProvider")
        print("CPUExecutionProvider passed")
    except Exception as e:
        print("CPUExecutionProvider failed")
        print(e)
    try:
        test_model(config, provider="CUDAExecutionProvider")
        print("CUDAExecutionProvider passed")
    except Exception as e:
        print("CUDAExecutionProvider failed")
        print(e)
```

Then I test the model on a simple input image on CPU and CUDA:

```
import onnxruntime
import numpy as np
import cv2

DATA_DIR = "data"


class MobileSAM_endtoend:
    def __init__(
        self, model_path, providers=["CUDAExecutionProvider", "CPUExecutionProvider"]
    ):
        # Initialize ONNX Runtime sessions
        self.model = onnxruntime.InferenceSession(model_path, providers=providers)
        # print input and outputs for reference
        print([(i.name, i.shape, i.type) for i in self.model.get_inputs()])
        print([(i.name, i.shape, i.type) for i in self.model.get_outputs()])

    def preprocess(self, image, bounding_box):
        height, width = image.shape[:2]
        resize_factor = max([width / 1024, height / 720])
        new_width = int(width / resize_factor)
        new_height = int(height / resize_factor)
        resized_size = (new_width, new_height)

        # Resize image
        image = cv2.resize(image, resized_size, interpolation=cv2.INTER_AREA)

        # pad to 720,1024
        image = np.pad(
            image,
            ((0, 720 - new_height), (0, 1024 - new_width), (0, 0)),
            mode="constant",
        )

        # calculate new box coordinates and return top-left and bottom-right coordinates
        coords = bounding_box.copy().reshape(2, 2)
        coords[..., 0] = coords[..., 0] / resize_factor
        coords[..., 1] = coords[..., 1] / resize_factor
        x1, y1, x2, y2 = coords.reshape(4)
        coords = np.array([[x1, y1], [x2, y2]], dtype=np.float16)
        labels = np.array(
            [2, 3], dtype=np.float16
        )  # 0: background point, 1: foreground point, 2: top-left box corner, 3: bottom-right box corner

        # add batch dimension
        image = image[None]
        coords = coords[None]
        labels = labels[None]

        return {
            "input": image,
            "point_coords": coords,
            "point_labels": labels,
            "mask_input": np.zeros((1, 1, 256, 256), dtype=np.float16),
            "has_mask_input": np.array([0], dtype=np.float16),
            "orig_im_size": np.array((720, 1024), dtype=np.float16),
        }, resized_size

    def get_foreground_mask(self, input_image, input_box):
        # input_image = cv2.cvtColor(cv2.imread(img_path), cv2.COLOR_BGR2RGB)
        inputs, resized_size = self.preprocess(input_image, input_box)

        foreground_mask = self.model.run(None, inputs)

        resized_width, resized_height = resized_size
        unpadded_mask = foreground_mask[0][0][:resized_height, :resized_width]
        mask = cv2.resize(
            unpadded_mask,
            input_image.shape[1::-1],
            interpolation=cv2.INTER_AREA,
        )
        return mask

    def __call__(self, img_path, input_boxes):
        return self.get_masks(img_path, input_boxes)


# for bug report
if __name__ == "__main__":
    model_path = "mobilesam/weights/SegmentAnythingBoundingBox.1.onnx"

    # test image: white circle on black background
    test_image = np.zeros((1024, 720, 3), dtype=np.uint8)
    test_image = cv2.circle(test_image, (512, 360), 100, (255, 255, 255), -1)

    # test bounding box: around the circle (412, 260, 612, 460)
    bounding_box = np.array([[412, 260, 612, 460]])

    # CPU
    sam = MobileSAM_endtoend(model_path, providers=["CPUExecutionProvider"])
    mask = sam.get_foreground_mask(test_image, bounding_box)
    print(f"CPU mask sum: {mask.sum()}")

    # GPU
    sam = MobileSAM_endtoend(model_path, providers=["CUDAExecutionProvider"])
    mask = sam.get_foreground_mask(test_image, bounding_box)
    print(f"GPU mask sum: {mask.sum()}")
```

Output below, as you can see when running on GPU I get a warning and the final output of the model is all zeros. 

> 2025-03-27 09:09:33.505742966 [W:onnxruntime:, constant_folding.cc:269 ApplyImpl] Could not find a CPU kernel and hence can't constant fold MatMul node '/pe_layer/MatMul'
> 2025-03-27 09:09:33.544000114 [W:onnxruntime:, constant_folding.cc:269 ApplyImpl] Could not find a CPU kernel and hence can't constant fold MatMul node '/pe_layer/MatMul'
> [('input', [1, 720, 1024, 3], 'tensor(uint8)'), ('point_coords', [1, 'num_points', 2], 'tensor(float16)'), ('point_labels', [1, 'num_points'], 'tensor(float16)'), ('mask_input', [1, 1, 256, 256], 'tensor(float16)'), ('has_mask_input', [1], 'tensor(float16)'), ('orig_im_size', [2], 'tensor(float16)')]
> [('uint8_mask', ['unk__809', 'unk__811', 'unk__812'], 'tensor(uint8)')]
> CPU mask sum: 7928030
> 2025-03-27 09:09:36.534940481 [W:onnxruntime:, constant_folding.cc:269 ApplyImpl] Could not find a CPU kernel and hence can't constant fold MatMul node '/pe_layer/MatMul'
> 2025-03-27 09:09:36.568173564 [W:onnxruntime:, constant_folding.cc:269 ApplyImpl] Could not find a CPU kernel and hence can't constant fold MatMul node '/pe_layer/MatMul'
> 2025-03-27 09:09:36.703409976 [W:onnxruntime:, transformer_memcpy.cc:83 ApplyImpl] 37 Memcpy nodes are added to the graph encoder_decoder_model for CUDAExecutionProvider. It might have negative impact on performance (including unable to run CUDA graph). Set session_options.log_severity_level=1 to see the detail logs before this message.
> 2025-03-27 09:09:36.710802614 [W:onnxruntime:, session_state.cc:1263 VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
> 2025-03-27 09:09:36.710823554 [W:onnxruntime:, session_state.cc:1265 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
> [('input', [1, 720, 1024, 3], 'tensor(uint8)'), ('point_coords', [1, 'num_points', 2], 'tensor(float16)'), ('point_labels', [1, 'num_points'], 'tensor(float16)'), ('mask_input', [1, 1, 256, 256], 'tensor(float16)'), ('has_mask_input', [1], 'tensor(float16)'), ('orig_im_size', [2], 'tensor(float16)')]
> [('uint8_mask', ['unk__809', 'unk__811', 'unk__812'], 'tensor(uint8)')]
> GPU mask sum: 0

P.S. Let me know if I should supply weights. Is not allowed by github to supply onnx, but I could change the extension and attach

### To reproduce

save create_onnx.py and bug_report.py with the code above

Install requirements:
git+https://github.com/ChaoningZhang/MobileSAM.git
onnxconverter-common
onnx
onnxruntime
onnxruntime-gpu
torch
segment-anything
matplotlib
numpy
opencv-python-headless
scikit-learn

Run the shell code
Run bug_report.py

### Platform

Linux

### OS Version

24.12.31

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

onnxruntime-gpu 1.21.0

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 12.2

**Comment by rene-wilbers-adesso** [comment]
Update: if I take out the conversion to float16 using the onnxconverter_common library, I get an ONNX that gives the same warning but the outputs are equal between CPU and GPU

**Comment by tianleiwu** [comment]
Indeed, CPU currently does not handle float16 for most operators, so CPU provider will inference in float32 for most cases(by casting inputs to float32 and outputs to float16), and GPU could handle float16 natively, that could cause difference when you compare float16 model result in CPU and GPU.

**Comment by rene-wilbers-adesso** [comment]
@tianleiwu actually I think it’s something else, because the output if the GPU is faulty. From CPU or MPS I get perfect results. CUDA is always empty mask full of zeros.

**Comment by tianleiwu** [comment]
@rene-wilbers-adesso, then the issue is regarding to float16 conversion causing accuracy loss.

**Comment by snnn** [comment]
This issue has been automatically closed because it has been marked as 'stale' for more than 30 days without activity. If you believe this is still an issue, please feel free to reopen it.
